### ISSUE TITLE
Feature request: JSON pre-processing

### ISSUE THREAD

**Issue by markspolakovs** [marcusolsson/grafana-json-datasource#207]
Firstly, thanks for writing and maintaining this plugin! It's been really useful

Often I've found myself writing nearly the same set of JSONPath/JSONata expressions for each field, with only minor differences. To take a trivial example, if I have an API that returns `{"array": [{}, {}, ...]}`, I need to prefix each field's query with `$.array[*]`.

It'd be useful if there was a way of applying a JSONPath/JSONata transformation to the complete response body before handing it off to the individual fields - in the above example this would involve a "global" expression of `$.array[*]`, and then each field could reference the array element individually.

I'd be up for writing up a PR to implement this if you think it'd be useful.

**Comment by marcusolsson** [comment]
Yes! This is something I've been thinking about for a while actually. Though from the perspective of "nesting" queries rather than "preprocessing" them.

Basically, you'd be able to create a child query from another query. Basically nesting them under each other.

![prototype](https://user-images.githubusercontent.com/8396880/140947832-6c420c86-2574-4765-a88b-19a40c161b42.png)

This would let you use both JSONata and JSONpath for different parts of the response.

Does that sound like what you were looking for?

**Comment by markspolakovs** [comment]
@marcusolsson Oh, I hadn't thought about it that way, but now that you mention it that would do what I want - I'd wrap all my field queries in one "group".

I'm curious about how you'd implement this - in my mind the fields would go from being a flat list to a "tree", where the root and branches are groups. Something like this:

```
┌─────────────┐
│  $.array[*] │
└┬────────────┘
 │
 ├─────── $.field1
 │
 │
 │        $.field2
 │
 │      ┌┬───────────────────┐
 └──────┼ $.anotherArray[*]  │
        └┼───────────────────┘
         │
         ├─────── $.subfield1
         │
         │
         └─────── $.subfield2
```

And then at runtime it'd traverse the tree, essentially flattening it to get the resulting fields, so you end up with just `[field1, field2, subfield1, subfield2]`. Have I understood your idea correctly?

**Comment by marcusolsson** [comment]
Exactly! It would essentially just be an easier way to build the queries. You'd still need each query to return arrays of the same length.

**Comment by markspolakovs** [comment]
Sounds good! I've started work on this on my fork at https://github.com/marcusolsson/grafana-json-datasource/compare/main...markspolakovs:marks/field-groups - currently I've only got the back-end (and it needs work), I'll submit a PR once it's ready.
